[Merged by Bors] - feat(tactic/rewrite_search): Automatically searching for chains of rewrites

lacker opened the pull request:
This pull request is based on a branch originally developed by @semorrison , @khoek , and @jcommelin . The idea of rewrite_search is a tactic that will search through chains of potential rewrites to prove the goal, when the goal is an equality or iff statement. There are three key components: `discovery.lean` finds a bunch of rules that can be used to generate rewrites, `search.lean` runs a breadth-first-search algorithm on the two sides of the quality to find a path that connects them, and `explain.lean` generates Lean code from the resulting proof, so that you can replace the call to `rewrite_search` with the explicit steps for it.

I removed some functionality from the rewrite_search branch and simplified the data structures somewhat in order to get this pull request small enough to be reviewed. If there is functionality from that branch that people particularly wanted, let me know and I can either include it in this PR or in a subsequent one. In particular, most of the configuration options are omitted.

For data structures, the whole `table` data structure is gone, replaced by a `buffer` and `rb_map` for efficient lookup. Write access to the buffer is also append-only for efficiency. This seems to be a lot faster, although I haven't created specific performance benchmarks.

kim-em committed:
feat(array/defs): functions to copy intro arrays

kim-em committed:
feat(array/defs): functions to copy intro arrays

kim-em committed:
feat(data/array/table): an API around array

kim-em committed:
feat(data/array/table): an API around array

kim-em committed:
formatting

kim-em committed:
formatting

khoek committed:
Some cleanup

khoek committed:
Some cleanup

khoek committed:
It compiles

khoek committed:
All tests pass

khoek committed:
lib: expr

khoek committed:
feat(tactic/expr): some expr utilities

khoek committed:
lib: tactic

khoek committed:
lib: array

khoek committed:
lib: table

khoek committed:
remove 'lib/num' since the graph tracer is gone

khoek committed:
space

khoek committed:
Some instances

khoek committed:
Some instances

khoek committed:
Some newlines

khoek committed:
Add reducible attribute

khoek committed:
Separate tests of discovery

khoek committed:
feat: the example bundles

khoek committed:
Better organise imports

khoek committed:
feat: `suggestion` command

khoek committed:
feat: the discovery subsystem

khoek committed:
lib: string

khoek committed:
lib: string

khoek committed:
Remove the last references to lib stuff

khoek committed:
type

khoek committed:
Remove now-unused config options

khoek committed:
better names

kim-em committed:
Merge branch 'master' into array

kim-em committed:
merge

kim-em committed:
Merge branch 'array' into rewrite-search

kim-em committed:
merge

kim-em committed:
merge

kim-em committed:
fix merge

jcommelin committed:
Merge branch 'master' into rewrite-search

jcommelin committed:
Document array.defs

jcommelin committed:
Golf the code

jcommelin committed:
Rename table_ref.from_nat to table_ref.of_nat

jcommelin committed:
Merge branch 'master' into rewrite-search

jcommelin committed:
Inline the function is_eq_or_iff_after_binders

jcommelin committed:
All import statements use full paths

jcommelin committed:
Oops, forgot an import statement

jcommelin committed:
Docstrings for 2nd half of core.common

jcommelin committed:
Fix another import statement

jcommelin committed:
Little stubs of documentation

jcommelin committed:
Question in a comment

jcommelin committed:
Comment out imports of data.rat

kim-em committed:
Merge remote-tracking branch 'origin/master' into rewrite-search

kim-em committed:
Merge remote-tracking branch 'origin/master' into rewrite-search

kim-em committed:
merge

kim-em committed:
Trying to get things working again

kim-em committed:
err, it's running again, but the explanations it gives are barking mad

kim-em committed:
oops!

kim-em committed:
merge

lacker committed:
clean up test

lacker committed:
remove some experimental-seeming tests

lacker committed:
remove simp stuff

lacker committed:
switch default to bfs

lacker committed:
remove pexplore

lacker committed:
trim unused

lacker committed:
switch default metric to trivial

lacker committed:
Remove edit distance

lacker committed:
remove a lot of tracer references

lacker committed:
a lot of long lines

lacker committed:
eliminate cutesey name

lacker committed:
remove tr from the search state

lacker committed:
pipe back unused tracer

lacker committed:
more rem unused

lacker committed:
lint errors

lacker committed:
fix lint

lacker committed:
fix tests

lacker committed:
Remove view

lacker committed:
remove tracer

lacker committed:
dont update the metric

lacker committed:
rm step args

lacker committed:
remove some unused methods, fix misspelled word

lacker committed:
remove a bunch of metrics

lacker committed:
remove metric

lacker committed:
flatten strategy dir

lacker committed:
remove unused estimate table

lacker committed:
yet more unused fns

lacker committed:
remove 3/4 of state class parametrization

lacker committed:
remove alpha-parametrization

lacker committed:
move bfs into engine

lacker committed:
get rid of the inst

lacker committed:
remove strategy

lacker committed:
stop having both init and startup

lacker committed:
remove dnum

lacker committed:
remove init_result stuff

lacker committed:
unused stats

lacker committed:
constructor cleanup

lacker committed:
flatten dir structure

lacker committed:
Remove separate rw config object

lacker committed:
remove some unused stuff

lacker committed:
remove unused table stuff

lacker committed:
remove concept of table_ref

lacker committed:
remove table alloc

lacker committed:
replace one table with a buffer

lacker committed:
use a buffer for rws

lacker committed:
replace another table with a buffer

lacker committed:
store tokens in a buffer

lacker committed:
eliminate vertex table

lacker committed:
unpoint to table

lacker committed:
remove unused array stuff

lacker committed:
unneeded indexed/keyed stuff

lacker committed:
remove stuff we dont use

lacker committed:
consolidate the tactic.lean stuff into interactive.lean

lacker committed:
unused tactic.lean

lacker committed:
consolidate hook and types

lacker committed:
consolidate common

lacker committed:
remove unused

lacker committed:
consolidate module

lacker committed:
Remove the concept of tokens

lacker committed:
remove sided pair

lacker committed:
flattening namespacse

lacker committed:
use std trace mechanism

lacker committed:
remove ability to choose nonoptimal backtracking

lacker committed:
make a lot of things private

lacker committed:
refactor to put more discovery stuff in discovery

lacker committed:
exhaustive config

lacker committed:
stop using rewrite_progress

lacker committed:
small removel

lacker committed:
Fix comp

lacker committed:
stop doubule bfsing

lacker committed:
rm curr depth

lacker committed:
remove bfs_state

lacker committed:
no need to return an exhausted iter

lacker committed:
some refactoring

lacker committed:
remove rewrite_iter

lacker committed:
flatten a namespace

lacker committed:
simplify

lacker committed:
no need to double filter on visited bool

lacker committed:
remove rws field from vertex

lacker committed:
more unused stuff

lacker committed:
rename

lacker committed:
cleanup

lacker committed:
remove old way of mutating vertices

lacker committed:
consolidate into engine

lacker committed:
rm

lacker committed:
clean up file naming

lacker committed:
redundancy, unused stuff

lacker committed:
search things in search

lacker committed:
clean up vertex stuff

lacker committed:
use rb_map to avoid quadratic blwup

lacker committed:
consolidation

lacker committed:
privacy accurate in search.lean

lacker committed:
misc

lacker committed:
clean up redundant test

lacker committed:
get rid of the joining-then-chopping pattern

lacker committed:
partway through

lacker committed:
search cleanup ok

lacker committed:
indenting

lacker committed:
comments on search.lean

lacker committed:
more comments

lacker committed:
rm one unused method

lacker committed:
comments for frontend

lacker committed:
fix up docstrings

lacker committed:
finish disco docsg

lacker committed:
docs

lacker committed:
remove some unused stuff

lacker committed:
fix lint

jcommelin commented on src/tactic/rewrite_search/discovery.lean:
```diff
@@ -0,0 +1,102 @@
+used by rewrite_search.
+-/
+private meta def rules_from_rw_rules (rws : list rw_rule) : tactic (list (expr × bool)) :=
+  rws.mmap (λ r, do e ← to_expr' r.rule, pure (e, r.symm))
```
```suggestion
rws.mmap (λ r, do e ← to_expr' r.rule, pure (e, r.symm))
```

lacker replied:
OK

jcommelin commented on src/tactic/rewrite_search/discovery.lean:
```diff
@@ -0,0 +1,102 @@
+meta def rewrite_search_attr : user_attribute := {
+  name := `rewrite,
+  descr := "declare that this definition should be considered by `rewrite_search`",
+}
```
```suggestion
meta def rewrite_search_attr : user_attribute :=
{ name := `rewrite,
  descr := "declare that this definition should be considered by `rewrite_search`" }
```

lacker replied:
OK

jcommelin commented on src/tactic/rewrite_search/discovery.lean:
```diff
@@ -0,0 +1,102 @@
+private meta def rules_from_rewrite_attr : tactic (list (expr × bool)) :=
+do names ← attribute.get_instances `rewrite,
+exprs ← names.mmap mk_const,
+return $ rules_from_exprs exprs
```
```suggestion
  exprs ← names.mmap mk_const,
  return $ rules_from_exprs exprs
```

lacker replied:
OK

jcommelin commented on src/tactic/rewrite_search/discovery.lean:
```diff
@@ -0,0 +1,102 @@
+private meta def from_tracked (rule_index : ℕ) (tracked : ℕ × tracked_rewrite) : rewrite :=
+do let (rw_index, rw) := tracked,
+let h : how := ⟨rule_index, rw_index, rw.addr⟩,
+⟨rw.exp, rw.proof, h⟩
```
```suggestion
  let h : how := ⟨rule_index, rw_index, rw.addr⟩,
  ⟨rw.exp, rw.proof, h⟩
```

lacker replied:
OK

jcommelin commented on src/tactic/rewrite_search/discovery.lean:
```diff
@@ -0,0 +1,102 @@
+meta def get_rewrites (rules : list (expr × bool)) (exp : expr) (cfg : config) :
+tactic (buffer rewrite) :=
+do lists ← list.mmap (rewrites_for_rule exp cfg) rules.enum,
+return (list.foldl buffer.append_list buffer.nil lists)
```
```suggestion
  tactic (buffer rewrite) :=
do lists ← list.mmap (rewrites_for_rule exp cfg) rules.enum,
  return (list.foldl buffer.append_list buffer.nil lists)
```

lacker replied:
OK

jcommelin commented on src/tactic/rewrite_search/explain.lean:
```diff
@@ -0,0 +1,255 @@
+/-- Convert the pair to a readable string format. -/
+def to_string [has_to_string α] (p : dir_pair α) : string :=
+  to_string p.l ++ "-" ++ to_string p.r
+instance has_to_string [has_to_string α] : has_to_string (dir_pair α) := ⟨to_string⟩
```
```suggestion
to_string p.l ++ "-" ++ to_string p.r

instance has_to_string [has_to_string α] : has_to_string (dir_pair α) := ⟨to_string⟩
```

lacker replied:
OK

jcommelin commented on src/tactic/rewrite_search/explain.lean:
```diff
@@ -0,0 +1,255 @@
+
+/- Convert a rule into the string of Lean code used to refer to this rule. -/
+private meta def pp_rule (r : expr × bool) : tactic string :=
+  do pp ← pp r.1, return $ (if r.2 then "←" else "") ++ (to_string pp)
```
```suggestion
do pp ← pp r.1, return $ (if r.2 then "←" else "") ++ (to_string pp)
```

lacker replied:
OK

jcommelin commented on src/tactic/rewrite_search/explain.lean:
```diff
@@ -0,0 +1,255 @@
+private meta def using_location.explain_rewrites (rs : list (expr × bool)) (s : side)
+(steps : list how) : tactic string := do
+  rules ← steps.mmap $ λ h : how, option.to_list <$> explain_using_location rs s h,
+  return $ string.intercalate ",\n" rules.join
```
```suggestion
  (steps : list how) : tactic string :=
do rules ← steps.mmap $ λ h : how, option.to_list <$> explain_using_location rs s h,
  return $ string.intercalate ",\n" rules.join
```

lacker replied:
OK

jcommelin commented on src/tactic/rewrite_search/explain.lean:
```diff
@@ -0,0 +1,255 @@
+/-- A data structure for the result of a splice operation. -/
+@[derive inhabited]
+inductive splice_result
+-- There was more of the addr to be added left, but we hit a rw
```
I think it makes sense to add these comments to the docstring, so that it's easy to read up on them in the hover text.

lacker replied:
OK

jcommelin commented on src/tactic/rewrite_search/explain.lean:
```diff
@@ -0,0 +1,255 @@
+/- Attempt to add new rewrites into the `app_addr` tree. -/
+private meta def splice_in (a : option app_addr) (rws : list ℕ) (s : list expr_lens.dir) :
+tactic splice_result :=
+  splice_in_aux rws a <$> to_congr_form s
```
```suggestion
  tactic splice_result :=
splice_in_aux rws a <$> to_congr_form s
```

lacker replied:
OK

jcommelin commented on src/tactic/rewrite_search/explain.lean:
```diff
@@ -0,0 +1,255 @@
+/- Construct a single `erw` tactic for the given rules. -/
+private meta def build_rw_tactic (rs : list (expr × bool)) (hs : list ℕ) : tactic string := do
+  rws ← (hs.map $ nth_rule rs).mmap pp_rule,
+  return $ "erw [" ++ (string.intercalate ", " rws) ++ "]"
```
```suggestion
private meta def build_rw_tactic (rs : list (expr × bool)) (hs : list ℕ) : tactic string :=
do rws ← (hs.map $ nth_rule rs).mmap pp_rule,
  return $ "erw [" ++ (string.intercalate ", " rws) ++ "]"
```

lacker replied:
OK

jcommelin commented on src/tactic/rewrite_search/explain.lean:
```diff
@@ -0,0 +1,255 @@
+
+/- Construct a string of Lean code that does a rewrite for the provided tree. -/
+private meta def explain_tree (rs : list (expr × bool)) (tree : app_addr) :
+tactic (list string) :=
```
```suggestion
  tactic (list string) :=
```

lacker replied:
OK

jcommelin commented on src/tactic/rewrite_search/explain.lean:
```diff
@@ -0,0 +1,255 @@
+/-- Explain a list of rewrites using `conv_x` tactics. -/
+meta def explain_rewrites (rs : list (expr × bool)) (s : side) (hows : list how) :
+tactic string :=
+  string.intercalate ",\n" <$> explanation_lines rs s none hows
```
```suggestion
  tactic string :=
string.intercalate ",\n" <$> explanation_lines rs s none hows
```

lacker replied:
OK

jcommelin commented on src/tactic/rewrite_search/explain.lean:
```diff
@@ -0,0 +1,255 @@
+private meta def explain_rewrites_concisely (steps : list (expr × bool)) (needs_refl : bool) :
+tactic string := do
+  rules ← string.intercalate ", " <$> steps.mmap pp_rule,
+  return $ "erw [" ++ rules ++ "]" ++ (if needs_refl then ", refl" else "")
```
```suggestion
  tactic string :=
do rules ← string.intercalate ", " <$> steps.mmap pp_rule,
  return $ "erw [" ++ rules ++ "]" ++ (if needs_refl then ", refl" else "")
```

lacker replied:
OK

jcommelin commented on src/tactic/rewrite_search/explain.lean:
```diff
@@ -0,0 +1,255 @@
+private meta def check_if_simple_rewrite_succeeds (rewrites : list (expr × bool)) (goal : expr) :
+tactic bool :=
+lock_tactic_state $ do
+  m ← mk_meta_var goal,
```
```suggestion
  tactic bool :=
lock_tactic_state $ do
  m ← mk_meta_var goal,
```

lacker replied:
OK

jcommelin commented on src/tactic/rewrite_search/explain.lean:
```diff
@@ -0,0 +1,255 @@
+
+/-- Construct a list of rewrites from a proof unit. -/
+meta def proof_unit.rewrites (u : proof_unit) (rs : list (expr × bool)) : list (expr × bool) :=
+  u.steps.filter_map $ how.to_rewrite rs
```
```suggestion
u.steps.filter_map $ how.to_rewrite rs
```

lacker replied:
OK

jcommelin commented on src/tactic/rewrite_search/explain.lean:
```diff
@@ -0,0 +1,255 @@
+  if explain_using_conv then
+    using_conv.explain_rewrites rs u.side u.steps
+  else
+    using_location.explain_rewrites rs u.side u.steps
```
```suggestion
  (explain_using_conv : bool) : tactic string :=
if explain_using_conv then
  using_conv.explain_rewrites rs u.side u.steps
else
  using_location.explain_rewrites rs u.side u.steps
```

lacker replied:
OK

jcommelin commented on src/tactic/rewrite_search/explain.lean:
```diff
@@ -0,0 +1,255 @@
+  head ← if rest.length = 0 ∨ u.side = side.L then pure [] else (do
+    n ← infer_type u.proof >>= (λ e, prod.snd <$> split_equation e) >>= pp,
+    pure $ ["transitivity " ++ to_string n]
+  ),
```
```suggestion
    pure $ ["transitivity " ++ to_string n]),
```

lacker replied:
OK

jcommelin commented on src/tactic/rewrite_search/explain.lean:
```diff
@@ -0,0 +1,255 @@
+    n ← infer_type u.proof >>= (λ e, prod.snd <$> split_equation e) >>= pp,
+    pure $ ["transitivity " ++ to_string n]
+  ),
+
```
Is there a reason you would like to keep this empty line?
```suggestion
```

lacker replied:
I don't mind removing it. done

jcommelin commented on src/tactic/rewrite_search/explain.lean:
```diff
@@ -0,0 +1,255 @@
+  return $ string.intercalate ",\n" expls
+
+private meta def explain_proof_concisely (rules : list (expr × bool)) (proof : expr)
+(l : list proof_unit) : tactic string :=
```
```suggestion
  (l : list proof_unit) : tactic string :=
```

lacker replied:
OK

jcommelin commented on src/tactic/rewrite_search/explain.lean:
```diff
@@ -0,0 +1,255 @@
+do let rws : list (expr × bool) := list.join $ l.map (λ u, do
+    (r, s) ← u.rewrites rules,
+    return (r, if u.side = side.L then s else ¬s)
+  ),
```
```suggestion
    return (r, if u.side = side.L then s else ¬s)),
```

lacker replied:
OK

jcommelin commented on src/tactic/rewrite_search/explain.lean:
```diff
@@ -0,0 +1,255 @@
+-/
+meta def explain_search_result (cfg : config) (rules : list (expr × bool)) (proof : expr)
+(units : list proof_unit) : tactic string := do
+  explanation ← explain_proof_concisely rules proof units <|>
```
```suggestion
  (units : list proof_unit) : tactic string :=
do explanation ← explain_proof_concisely rules proof units <|>
```

lacker replied:
OK

jcommelin commented on src/tactic/rewrite_search/frontend.lean:
```diff
@@ -0,0 +1,62 @@
+current target, and generates a string explanation for it.
+-/
+private meta def rewrite_search_target (cfg : config) (rws : list rw_rule) :
+tactic string :=
```
```suggestion
  tactic string :=
```

lacker replied:
OK

jcommelin commented on src/tactic/rewrite_search/search.lean:
```diff
@@ -0,0 +1,208 @@
+  path2 ← walk_up_parents g v.parent,
+  match v.side with
+    | side.L := return (path2.reverse, path1.reverse)
+    | side.R := return (path1.reverse, path2.reverse)
```
```suggestion
  | side.L := return (path2.reverse, path1.reverse)
  | side.R := return (path1.reverse, path2.reverse)
```

lacker replied:
OK

jcommelin commented on src/tactic/rewrite_search/search.lean:
```diff
@@ -0,0 +1,208 @@
+    let new_vertex : vertex := ⟨new_vertex_id, rw.exp, pp, v.side, (some new_edge)⟩,
+    trace_if_enabled `rewrite_search format!"new edge: {v.pp} → {new_vertex.pp}",
+    return { g with vertices := g.vertices.push_back new_vertex,
+                    vmap := g.vmap.insert pp new_vertex_id }
```
```suggestion
| (some id) := do
  existing_vertex ← g.get_vertex id,
  if v.side = existing_vertex.side then return g
  else return { g with solving_edge := some ⟨v.id, existing_vertex.id, rw.proof, rw.how⟩ }
| none := do
  let new_vertex_id := g.vertices.size,
  let new_edge : edge := ⟨v.id, new_vertex_id, rw.proof, rw.how⟩,
  let new_vertex : vertex := ⟨new_vertex_id, rw.exp, pp, v.side, (some new_edge)⟩,
  trace_if_enabled `rewrite_search format!"new edge: {v.pp} → {new_vertex.pp}",
  return { g with vertices := g.vertices.push_back new_vertex,
                  vmap := g.vmap.insert pp new_vertex_id }
```

lacker replied:
I aligned these by indenting the "let" and "match" instead, I think that is better?

jcommelin replied:
Ooh, you're right. Thanks!

jcommelin commented on src/tactic/rewrite_search/search.lean:
```diff
@@ -0,0 +1,208 @@
+-/
+private meta def expand_vertex (v : vertex) : tactic graph :=
+do rws ← get_rewrites g.rules v.exp g.conf,
+list.mfoldl (λ g rw, add_rewrite g v rw) g rws.to_list
```
```suggestion
  list.mfoldl (λ g rw, add_rewrite g v rw) g rws.to_list
```

lacker replied:
OK

jcommelin commented on src/tactic/rewrite_search/search.lean:
```diff
@@ -0,0 +1,208 @@
+  g ← expand_vertex g v,
+  match g.solving_edge with
+    | some _ := return g
+    | none   := find_solving_edge g (vertex_id + 1)
```
```suggestion
do let v := g.vertices.read (fin.mk vertex_id h),
  g ← expand_vertex g v,
  match g.solving_edge with
  | some _ := return g
  | none   := find_solving_edge g (vertex_id + 1)
```

lacker replied:
OK

jcommelin commented on src/tactic/rewrite_search/search.lean:
```diff
@@ -0,0 +1,208 @@
+match proofs with
+  | []              := fail "cannot combine empty proof list"
+  | (proof :: rest) := list.mfoldl mk_eq_trans proof rest
+end
```
```suggestion
private meta def combine_proofs : list expr → tactic expr :=
| []              := fail "cannot combine empty proof list"
| (proof :: rest) := list.mfoldl mk_eq_trans proof rest
```

lacker replied:
OK

jcommelin commented on src/tactic/rewrite_search/search.lean:
```diff
@@ -0,0 +1,208 @@
+  proofs ← match s with
+    | side.L := edges.mmap (λ e, e.proof)
+    | side.R := edges.reverse.mmap (λ e, e.proof >>= mk_eq_symm)
+  end,
```
```suggestion
    end,
```

lacker replied:
OK

jcommelin reviewed (commented):
Thanks a lot for reviving this! I left a ton of cosmetic comments, but someone who actually understands tactic writing should do a review of the contents.

bryangingechen commented on src/tactic/rewrite_search/discovery.lean:
```diff
@@ -0,0 +1,102 @@
+
+open tactic tactic.interactive tactic.rewrite_search
+
+/-
```
Here and below, doc string comments should start with `/--` in order to show up when you hover over them and for them to show up on the website.
```suggestion
/--
```

lacker replied:
The private methods should not be used by anyone outside this file (well, if I understand Lean method visibility correctly they *cannot* be used by anyone outside this file), so I think they should not have docstrings that should show up on the website. This is the rule that the linter enforces, too, that private methods don't require docstrings, & this does pass the lint checks. I added regular comments so that there is still an explanation of what they do, for users who are in the same file they are defined.

bryangingechen replied:
Turning the comments into docstrings will make things more convenient for people working in that file by letting them see the comments simply by hovering over the declaration names, rather than scrolling up. Also, the declarations of the private functions are still going to show up in the online docs, and I see no reason why the comments about them shouldn't appear with them as well.

lacker replied:
All right, I changed them to docstrings.

lacker committed:
respond to feedbacks

lacker committed:
more stylistic fixes

lacker committed:
overall documentation

lacker committed:
add reference

Vierkantor commented on src/tactic/rewrite_search/discovery.lean:
```diff
@@ -0,0 +1,99 @@
+once in each direction.
+-/
+private meta def rules_from_exprs (l : list expr) : list (expr × bool) :=
+  l.map (λ e, (e, ff)) ++ l.map (λ e, (e, tt))
```
```suggestion
l.map (λ e, (e, ff)) ++ l.map (λ e, (e, tt))
```

lacker replied:
OK

Vierkantor commented on src/tactic/rewrite_search/discovery.lean:
```diff
@@ -0,0 +1,99 @@
+
+/-- Collect all hypotheses in the local context that are usable as rewrite rules. -/
+private meta def rules_from_hyps : tactic (list (expr × bool)) := do
+  hyps ← local_context,
```
This seems to be the convention used throughout the file:
```suggestion
private meta def is_acceptable_hyp (r : expr) : tactic bool :=
do t ← infer_type r >>= whnf, return $ is_acceptable_rewrite t ∧ ¬t.has_meta_var

/-- Collect all hypotheses in the local context that are usable as rewrite rules. -/
private meta def rules_from_hyps : tactic (list (expr × bool)) :=
do hyps ← local_context,
```

lacker replied:
OK

Vierkantor commented on src/tactic/rewrite_search/explain.lean:
```diff
@@ -0,0 +1,255 @@
+open splice_result
+
+private meta def pack_splice_result (s : expr_lens.dir) :
+splice_result → dir_pair (option app_addr) → splice_result
```
```suggestion
  splice_result → dir_pair (option app_addr) → splice_result
```

lacker replied:
OK

Vierkantor commented on src/tactic/rewrite_search/explain.lean:
```diff
@@ -0,0 +1,255 @@
+| sr _ := sr
+
+private meta def splice_in_aux (new_rws : list ℕ) :
+option app_addr → list expr_lens.dir → splice_result
```
```suggestion
  option app_addr → list expr_lens.dir → splice_result
```

lacker replied:
OK

Vierkantor commented on src/tactic/rewrite_search/explain.lean:
```diff
@@ -0,0 +1,255 @@
+
+/-- Explain a single rewrite using `nth_rewrite`. -/
+private meta def explain_using_location (rs : list (expr × bool)) (s : side) :
+how → tactic (option string)
```
```suggestion
  how → tactic (option string)
```

lacker replied:
OK

Vierkantor commented on src/tactic/rewrite_search/frontend.lean:
```diff
@@ -0,0 +1,100 @@
+Search for a chain of rewrites to prove an equation or iff statement.
+Includes the rewrite rules specified in the same way as the `rw` tactic accepts.
+-/
+meta def rewrite_search_with (try_harder : parse $ optional (tk "!")) (rs : parse rw_rules)
```
Wouldn't it be nicer to match the syntax of `simp`, by having `rs` be an optional argument? So `by { rewrite_search }` works the same as it does now, and `by { rewrite_search [h1, h2, h3] }` is what is now `by { rewrite_search_with [h1, h2, h3] }`.

lacker replied:
That's a good idea, I'll take a stab at it.

lacker replied:
OK, this is in now. The code is a decent amount simpler. Many thanks to Rob's advice in Zulip ;-)

Vierkantor commented on src/tactic/rewrite_search/frontend.lean:
```diff
@@ -0,0 +1,100 @@
+  rules ← collect_rules rws,
+  g ← mk_graph cfg rules t,
+  (_, proof, steps) ← g.find_proof,
+  tactic.exact proof >> (explain_search_result cfg rules proof steps)
```
Why not:
```suggestion
  tactic.exact proof,
  explain_search_result cfg rules proof steps
```

lacker replied:
OK

Vierkantor commented on src/tactic/rewrite_search/search.lean:
```diff
@@ -0,0 +1,208 @@
+
+/--
+Construct a graph to search for a proof of a given equation.
+-/
```
Did I understand this correctly?
```suggestion
/--
Construct a graph to search for a proof of a given equation.

This graph will initially only contain two disconnected vertices corresponding to the two sides of the equation.
As we search in the graph, we automatically add new vertices and edges.
-/
```

lacker replied:
Yes, I expanded the comment to this effect.

Vierkantor commented on src/tactic/rewrite_search/search.lean:
```diff
@@ -0,0 +1,208 @@
+-/
+private meta def add_rewrite (v : vertex) (rw : rewrite) : tactic graph :=
+do pp ← to_string <$> tactic.pp rw.exp,
+  let maybe_id := g.vmap.find pp,
```
We seem to match the new expression only by string equality. How do you deal with the case that two expressions have the same `pp` but do not unify, which seems to return an invalid proof here? (There is also the converse: two expressions unify but do not have the same `pp`. Continuing the search would be fine in that case I assume.)

lacker replied:
Hmm, this is a good point that I didn't consider. I thought about ways to handle this, but I'm not quite sure what to do. You are correct that it's okay if this expression matching misses two expressions that unify, but it's bad if it accepts expressions that actually do not unify. I think the ideal way would be to implement hashing on expressions in such a way that syntactically equivalent expressions had the same hash, and collisions were very unlikely. But, I don't think we can do that efficiently in Lean right now. We could check for unification and carry on, but once two nodes are mistakenly identified with this algorithm, it seems like many nodes will be mistakenly unified. So... the best suggestion I have is just to fail in this case. I at least added a comment to this effect in the overall summary. In practice, I have not run into this in any of the cases I tested at least.

robertylewis replied:
Ah, I see my earlier comment comes up here. What's the problem with making `vmap` an `expr_map nat`, i.e. `rb_map expr nat`? Is the problem that you need to match things that are "pp-equal" but not indentical, e.g. they have different but defeq type class instances?

lacker replied:
Well, I wasn't even aware that it was possible to use an `expr` as the key of an `rb_map`. It is certainly aesthetically superior to use an `expr_map`. So I tried it out, and unfortunately it makes the algorithm about 10x slower on the `rewrite_search.lean` test file, from 2 second compilation to 20 seconds. This map and the buffer are just in the inner loop of this algorithm, and AFAICT doing the expression comparisons is slow. I think to make this work we need to be cranking out all the speed improvements we can get, and in practice `pp` equality seems to be good enough. But ideally this entire map could be a c++ data structure and the inner loop of this logic could be in c++, but I don't want to get that messy in a way that will likely be incompatible with Lean 4. I'd like to revisit this once Lean 4 is working with mathlib though.

Vierkantor reviewed (changes requested):
My first set of remarks, I didn't have time to look at the rest of the code. I'm happy with how it is looking so far!

lacker committed:
respond to comments

lacker committed:
Remove unused try_harder

lacker committed:
use := instead of dot

robertylewis commented on src/tactic/rewrite_search/frontend.lean:
```diff
@@ -0,0 +1,101 @@
+can be reversed.
+
+One drawback of this algorithm to note is that it uses `pp` of an expression to determine
+equality. So the `rewrite_search` tactic is likely to fail when there are two different
```
Why?? I remember Scott talking about this years ago, where he was using string edit distance to choose which rewrites to prefer. This sounded weird but motivated, at least. If you're just testing equality, why not just compare the expressions? Is it "syntactic equality is too strong, definitional equality is too expensive, pp equality is a tradeoff?"

lacker replied:
See other comment, this is necessary for speed, hopefully fixable in Lean 4.

lacker replied:
I updated this comment to include this explanation

robertylewis commented on src/tactic/rewrite_search/types.lean:
```diff
@@ -0,0 +1,82 @@
+
+meta instance side.has_to_format : has_to_format side := ⟨side.to_string⟩
+
+/-- `how` contains information needed by the explainer to generate code for a rewrite. -/
```
Could you describe what the fields represent? I'm guessing `rule_index` denotes which rule in the static list of rules is used, `location` describes which match of that rule was used (in `nth_rewrite` syntax). But then `addr` looks maybe the same as `location`?

robertylewis replied:
This wasn't resolved

lacker replied:
`addr` is telling you which subexpression in the tree the rewrite applies to. I added descriptions for all these. `nth_rewrite` basically uses `addr` as well internally and the `expr_lens` structure is from there.

robertylewis commented on src/tactic/rewrite_search/types.lean:
```diff
@@ -0,0 +1,82 @@
+`max_iterations` controls how many vertices are expanded in the graph search.
+`explain` generates Lean code to replace the call to `rewrite_search`.
+`explain_using_conv` changes the nature of the explanation.
+`inflate_rws` makes rewrite search consider more rewrites.
```
There's no `inflate_rws` field

lacker replied:
Fixed

robertylewis commented on src/tactic/rewrite_search/types.lean:
```diff
@@ -0,0 +1,82 @@
+/--
+Split an equation (or an iff) into its left and right parts.
+-/
+meta def split_equation : expr → tactic (expr × expr)
```
This can move to `tactic.core` and become `is_eq_or_iff`. At some point we should collect all of these scattered matching functions into a unified API with tactic and option versions, but not now.

lacker replied:
Hmm, I don't see other functions like this in `tactics.core`. Is that the file you meant? We don't just want to check if it is an equality, we want to split the two sides up, fwiw.

robertylewis replied:
Sorry, the name should be `tactic.match_eq_or_iff`, corresponding to `tactic.match_eq`, `tactic.match_iff`, etc. These are defined in core Lean but better to add this one to mathlib for now. `is_eq_or_iff` would be the `option`-valued version. You can just stick it at the bottom of `tactic.core`. 

Getting both sides of a binary function expr application is a pretty common thing to want. It doesn't require the tactic monad, just `option`, but it's usually done in tactics so those variants are useful. Because there's no centralized collection of all the various versions of these, they're scattered throughout the library and probably duplicated a bunch. Let's not contribute to the scattering more than we have to.

lacker replied:
Ah, I wasn't aware of `match_eq` and `match_iff`! It seems simpler to just use `match_eq x <|> match_iff x` than to write this new function at all. I just rewrote to use those and removed `split_equation`.

robertylewis commented on src/tactic/rewrite_search/discovery.lean:
```diff
@@ -0,0 +1,99 @@
+private meta def rules_from_rewrite_attr : tactic (list (expr × bool)) :=
+do names ← attribute.get_instances `rewrite,
+  exprs ← names.mmap mk_const,
+  return $ rules_from_exprs exprs
```
```suggestion
  rules_from_exprs <$> names.mmap mk_const
```

lacker replied:
OK

robertylewis commented on src/tactic/rewrite_search/search.lean:
```diff
@@ -0,0 +1,212 @@
+/--
+Add all single-step rewrites starting at a particular vertex to the graph.
+-/
+private meta def expand_vertex (v : vertex) : tactic graph :=
```
For the most part this `search` and `discovery` are nicely modularized. This is the only declaration that breaks if `search` doesn't import `discovery`. Is it refactorable? Partly just leaving this as a note for myself as I continue to review.

robertylewis replied:
Okay, after reading more, it makes more sense that `discovery` is a module used by the search.

robertylewis commented on src/tactic/rewrite_search/search.lean:
```diff
@@ -0,0 +1,212 @@
+`pp` is the string format of the expression; we store this in the vertex to avoid
+recalculating it.
+`side` is whether this vertex descends from the left or right side of the equation.
+`parent` is the edge that originally added this vertex to the graph.
```
```suggestion
* `id` is a numerical id used to refer to this vertex in the context of a single graph.
* `exp` is the expression this vertex represents.
* `pp` is the string format of the expression; we store this in the vertex to avoid
  recalculating it.
* `side` is whether this vertex descends from the left or right side of the equation.
* `parent` is the edge that originally added this vertex to the graph.
```

lacker replied:
OK

robertylewis commented on src/tactic/rewrite_search/search.lean:
```diff
@@ -0,0 +1,212 @@
+`vmap` maps vertex.pp to vertex.id, so we can quickly find collisions.
+`solving_edge` represents a solution that will prove our target equation.
+It is an edge that would connect the two trees, so `solving_edge.fr` and `solving_edge.to`
+are vertices in different trees.
```
```suggestion
* `conf` and `rules` determine what rewrites are used to generate new graph vertices.
  Here, the format of a rewrite rule is an expression for rewriting, plus a flag for the
  direction to apply it in.
* `vertices` maps vertex.id to vertex.
* `vmap` maps vertex.pp to vertex.id, so we can quickly find collisions.
* `solving_edge` represents a solution that will prove our target equation.
  It is an edge that would connect the two trees, so `solving_edge.fr` and `solving_edge.to`
  are vertices in different trees.
```

lacker replied:
OK

robertylewis commented on src/tactic/rewrite_search/search.lean:
```diff
@@ -0,0 +1,212 @@
+  return ⟨conf, rules, [lhs_vertex, rhs_vertex].to_buffer,
+          native.rb_map.of_list [(lhs_pp, 0), (rhs_pp, 1)], none⟩
+
+variables (g : graph)
```
I very much sense the presence of a state monad here. I won't insist on a refactor, but I suspect this would look really clean in that style.

lacker replied:
I don't understand this comment. I'm not saying it's wrong, I just don't understand. But I'm excited about the "won't insist on a refactor" part of your comment ;-)

robertylewis replied:
You can [extend a monad (e.g. `tactic`) with extra state](https://leanprover-community.github.io/mathlib_docs/init/control/state.html#state_t). In this case, you'd effectively add a `graph` object to the tactic state. So your functions would be of type `graph_tactic T` instead of `tactic T`, and wouldn't have to have a `graph` object as input or output, since you functionally mimic having a global `graph`.

robertylewis commented on src/tactic/rewrite_search/search.lean:
```diff
@@ -0,0 +1,212 @@
+/--
+A convenience wrapper around `g.vertices` that fails if the index is out of bounds.
+-/
+meta def get_vertex (i : ℕ) : tactic vertex :=
```
This is a generic buffer function, it could be moved to `data.buffer` and renamed (`buffer.read_t`? `buffer.read_safe`? Not sure we have a naming convention for this.)

lacker replied:
`buffer` isn't defined in `mathlib`, though, it's defined at https://github.com/leanprover-community/lean/tree/ed6accfb611bbd28ef5b7e34361d2b58a366e2c9/library/data/buffer . Let's just punt on this?

robertylewis replied:
Mathlib's [data.buffer.basic](https://github.com/leanprover-community/mathlib/blob/master/src/data/buffer/basic.lean) is a totally reasonable place for it. If it's not moved now I doubt it will be moved any time soon.

lacker replied:
That file appears to be unused, and when I add `import data.buffer.basic` my code no longer compiles.

lacker replied:
Tracked it down to naming a variable "to", renamed, now there is a `read_t` function.

robertylewis commented on src/tactic/rewrite_search/search.lean:
```diff
@@ -0,0 +1,212 @@
+/--
+Find a list of edges that connect the given edge to the root of its tree.
+The edges are returned in leaf-to-root order, while they are in root-to-leaf direction,
+so if you want them in the logical order you must reverse the returned list.
```
This function is used in exactly one place and it's immediately reversed, why not just return them in the logical order to begin with?

lacker replied:
The recursion uses them in this order, and it's O(n^2) to do a reverse or concat during the recursion.

robertylewis commented on src/tactic/rewrite_search/search.lean:
```diff
@@ -0,0 +1,212 @@
+/--
+Add the new vertex and edge to the graph, that can be proved in one step starting
+at a given vertex, with a given rewrite expression.
+For efficiency, it's important that this is the only way the graph is mutated,
```
Ah, maybe this is the reason not to use a monad. I remember some hell dealing with reference counting objects in the monad state.

lacker replied:
This is the reason to use a buffer, so you can be very careful about only appending to it and never mutating it, rather than passing around lists or arrays, where it's more convenient but slower to modify things in different locations. I'm not sure about the monad part.

robertylewis replied:
Right. In my experience a while back (granted, n=1) having a buffer in the monad state did not play well with the reference counting.

robertylewis commented on src/tactic/rewrite_search/explain.lean:
```diff
@@ -0,0 +1,255 @@
+do explanation ← explain_proof_concisely rules proof units <|>
+                 explain_proof_full rules cfg.explain_using_conv units,
+  if cfg.explain then trace $ "/- `rewrite_search` says -/\n" ++ explanation
+  else skip,
```
```suggestion
  when cfg.explain $ trace $ "/- `rewrite_search` says -/\n" ++ explanation,
```
Also, this should certainly be hooked up to the "Try this:" machinery. (See e.g. `squeeze_simp`)

lacker replied:
OK. TIL this existed, I've been a poor impoverished emacs user

robertylewis commented on src/tactic/rewrite_search/frontend.lean:
```diff
@@ -0,0 +1,101 @@
+  explain_search_result cfg rules proof steps
+
+/-- Search for a chain of rewrites to prove an equation or iff statement. -/
+meta def rewrite_search (cfg : config := {}) : tactic string :=
```
We should use syntax `rewrite_search?` for `explain := tt`. This matches `rcases?`, `tidy?`, etc.

lacker replied:
OK - yeah this is actually a lot more convenient

robertylewis reviewed (commented):
This looks very good overall. I haven't carefully reviewed `explain.lean` yet.

My biggest concern is the obvious one: I don't get how this can be effective at scale. The more `@[rewrite]` lemmas that get tagged, the slower it will become. It's doing an unconstrained, undirected exploding search. I'm okay with merging it to see how it goes, but consider my skepticism noted early!

I think we'll need a policy in mathlib that calls to `rewrite_search` must be replaced with the output explanation.

robertylewis commented on src/tactic/rewrite_search/frontend.lean:
```diff
@@ -0,0 +1,101 @@
+
+/-- Search for a chain of rewrites to prove an equation or iff statement. -/
+meta def rewrite_search (cfg : config := {}) : tactic string :=
+rewrite_search_target cfg []
```
```suggestion
meta def rewrite_search (cfg : config := {}) : tactic unit :=
rewrite_search_target cfg [] $> ()
```
Apparently it's not enforced that interactive tactics don't return data, but I think it's good style to preserve.

lacker replied:
OK - i cleaned up some of the other functions in the chain to not return strings either. I had some incorrect fuzzy idea that this string was used for something

robertylewis reviewed (commented):
Missed one

lacker committed:
respond to feedback

lacker committed:
rw_search_parser

lacker committed:
simplify after using custom parser

lacker committed:
use a question mark

lacker committed:
try this

lacker committed:
fix the defeq vs pp-eq issue

lacker committed:
improve comment

lacker committed:
fix comment

lacker committed:
read_t

lacker committed:
remove split_equation

robertylewis committed:
postpone elaboration until after parsing

During tactic parsing, local constants may not be in context, leading to possible failures like:

```
axiom bad {α : Type} [has_zero α] (x : α) : x = 0
example {α : Type} [has_zero α] (x : α) : x = 0 := by rewrite_search [bad x] -- fails: parsing time
```

lacker committed:
the empty rewrite list

lacker committed:
fix traces with explicit arguments

lacker committed:
add missing indents

Vierkantor reviewed (approved):
I haven't kept up with the changes since I made the suggestions, and from a quick read-through everything is looking good. I'll approve the PR and let another reviewer decide if it's mergeable.
